### PR TITLE
(feat) O3-3886: Demo Ward App in the Ref App

### PR DESCRIPTION
**Pull request by mogoodrich**

### CODE REVIEW

File: distro/configuration/locations/locations-core_demo.csv [hunk 1]
```diff
@@ -1,116 +1,13 @@
...
-8d243ead-936f-4bc8-b8a6-95baa2ed5217,,Ward 53,Ward 53,,Login Location,,,,,,,,,,,,,,,,
-9212bcda-32dc-4fb8-a517-8ade92908816,,Ward 54,Ward 54,,Login Location,,,,,,,,,,,,,,,,
-b8457ebe-48cd-4e6b-8139-23b269017d45,,Ward 55,Ward 55,,Login Location,,,,,,,,,,,,,,,,
-423ca603-a696-4a25-a00f-2e998df01a9f,,Ward 56,Ward 56,,Login Location,,,,,,,,,,,,,,,,
-9b35114a-a1cc-400c-ba1e-ee61538feeb7,,Ward 57,Ward 57,,Login Location,,,,,,,,,,,,,,,,
-aa661b2b-bf9f-4335-965c-9dcb5b6adf94,,Ward 58,Ward 58,,Login Location,,,,,,,,,,,,,,,,
-8874c0eb-be2b-4c9c-926c-bdd1e4394b99,,Ward 59,Ward 59,,Login Location,,,,,,,,,,,,,,,,
-2faca5a9-bfce-4cdc-8e09-690411230bcc,,Ward 60,Ward 60,,Login Location,,,,,,,,,,,,,,,,
\ No newline at end of file
+Uuid,Void/Retire,Name,Description,Parent,Tag|Login Location,Tag|Visit Location,Tag|Facility Location,Tag|Queue Location,Tag|Admission Location,Tag|Transfer Location,Attribute|9eca4f4e-707f-4bb8-8289-2f9b6e93803c,Attribute|Last Audit Date,Address 1,Address 2,Address 3,Address 4,Address 5,Address 6,City/Village,County/District,State/Province,Postal Code,Country,_order:1000
```
[the remark this answers is not in the document]
Comment (mogoodrich) [reply distro/configuration/locations/locations-core_demo.csv:1]: No problem at all @ibacher ... I though you mentioned on Slack (or somewhere else?) being good with doing away with them, should be find to add them back.

File: distro/configuration/dispositionConfig.json [hunk 2]
```diff
@@ -0,0 +1,26 @@
+[
+  {
+    "uuid": "66de7f60-b73a-11e2-9e96-0800200c9a66",
+    "name": "disposition.admit",
+    "type": "ADMIT",
+    "conceptCode": "PIH:ADMIT TO HOSPITAL",
```
[the remark this answers is not in the document]
Comment (mogoodrich) [reply distro/configuration/dispositionConfig.json:6]: I think we have figured it out, just waiting on the new CIEL release from Andy: https://talk.openmrs.org/t/concepts-needed-for-ipd-ward-app-in-o3-ref-app-distro/43805/9

File: distro/configuration/locations/locations-core_demo.csv (hunk 1 above)
[the remark this answers is not in the document]
Comment (ibacher) [reply distro/configuration/locations/locations-core_demo.csv:1]: Ah... I think I meant it's ok to get rid of the "Ward N" ones. I should've been more explicit.

File: distro/configuration/dispositionConfig.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (ibacher) [reply distro/configuration/dispositionConfig.json:6]: Yeah, saw that after I wrote this.

File: distro/configuration/encountertypes/encountertypes_core-demo.csv [hunk 3]
```diff
@@ -13,3 +13,6 @@ d7151f82-c1f3-4152-a605-2f9ea7414a79,,Visit Note,"Encounter where a full or abbr
 39da3525-afe4-45ff-8977-c53b7b359158,,Order,Encounter for capturing orders,,,
 29c02aff-9a93-46c9-bf6f-48b552fcb1fa,,Immunizations,Encounter for capturing immunizations,,,
 36db5123-0ad5-41c0-9037-625b46e0ceef,,Mental Health Assessment,Mental Health Assessment,,,
+d3b07384-8d1c-4e6b-9b8e-2f3b8e4a1c9f,,Transfer,Indicates that a patient is being transferred into a different department within the hospital. (Transfers out of the hospital should not use this encounter type.),,,
+a1f5c3d2-4b6e-4e8a-9f2d-1b3e8e4a2d7f,,Inpatient Note,A short free-text note written by a clinician about an inpatient during their stay,,,
+b2c4d5e6-7f8a-4e9b-8c1d-2e3f8e4a3b8f,,Transfer Request,A stand-alone encounter type for a transfer request not part of a larger note,,,
```
Comment (mogoodrich) [inline distro/configuration/encountertypes/encountertypes_core-demo.csv:18]: Adds some new encounter types associated with ADT events

File: distro/configuration/locations/locations-core_demo.csv [hunk 4]
```diff
@@ -1,116 +1,60 @@
...
+8d6c993e-c2cc-11de-8d13-0010c6dffd0f,,Unknown Location,Unknown Location,,,,,,,,,,,,,,,,,,,,,
```
Comment (mogoodrich) [inline distro/configuration/locations/locations-core_demo.csv:2]: The most noteworthy thing about the changes to locations is that it sets up a location hierarchy at "Ubuntu Hospital""

image

This hierarchy is necessary for the Ward module to work, and modelling general way EMR-API "views the world": when you have multiple locations within a single facility the sites should be arranged as children of a main site location, and the main location (in this case "Ubuntu Hospital" should be tagged as a visit location.  In order for the Inpatient functionality to work and to allow admission and transfer between different locations, a patient's active visit needs to be at the parent site.

Comment (mogoodrich) [reply distro/configuration/locations/locations-core_demo.csv:2]: (Note that I'm not sure what "Facility Location" means within this demo, so I left it as-is.

File: distro/pom.xml [hunk 5]
```diff
@@ -49,6 +49,9 @@
     <ordertemplates.version>1.0.2</ordertemplates.version>
     <patientflags.version>3.0.8</patientflags.version>
     <o3forms.version>2.3.0</o3forms.version>
+    <emrapi.version>2.1.0-SNAPSHOT</emrapi.version>
+    <event.version>2.10.0</event.version>
+    <bedmanagement.version>6.0.0</bedmanagement.version>
```
Comment (mogoodrich) [inline distro/pom.xml:54]: New modules added.... EMR API, Event and Bed management

**Comment by mogoodrich** [comment]
Moving this out of draft, as it is ready for review, see my comments!

Note we will need the following Iniz issue reviewed and merged before this will work:

https://github.com/mekomsolutions/openmrs-module-initializer/issues/284

**Comment by mogoodrich** [comment]
[Screencast from 2024-10-25 19-13-47.webm](https://github.com/user-attachments/assets/24dbf814-c454-4845-a091-cb649b9d26ad)

**Review by ibacher — APPROVED** [review]

**Comment by denniskigen** [comment]
Yes, @mogoodrich, you can download the artifact named `report patient management` on this page https://github.com/openmrs/openmrs-distro-referenceapplication/actions/runs/11526530278?pr=851

**Comment by denniskigen** [comment]
CleanShot 2024-10-31 at 9  34 17@2x

Looks like the issue is that the `provider.uuid` is blank here.

**Comment by mogoodrich** [comment]
> Dunno why the `provider.uuid` here is blank [here](https://github.com/openmrs/openmrs-esm-patient-management/blob/main/e2e/specs/active-visits.spec.ts#L28) when creating the encounter for running the test.

Yeah, I thought that was where the error was... yeah, no reason any of my changes should result in the provider not being avaiable?

**Comment by denniskigen** [comment]
Yeah...

**Comment by mogoodrich** [comment]
> Yeah...

But's pretty consistently failing... got drawn into other stuff today, hopefully I can take a look tomorrow and have a eureka... :)

**Comment by mogoodrich** [comment]
> But's pretty consistently failing... got drawn into other stuff today, hopefully I can take a look tomorrow and have a eureka... :)

Got it @denniskigen !  Need to dig a little deeper to see what exactly is going on and what the fix it... likely just a metadata tweak in the ref app data... but on startup the EMR API module makes sure there is an "Unknown" provider in the system, and in our case it's overwriting the "admin" provider set up by ref app. Will keep you posted, but thanks for the help!

**Comment by mogoodrich** [comment]
> Got it @denniskigen ! Need to dig a little deeper to see what exactly is going on and what the fix it... likely just a metadata tweak in the ref app data... but on startup the EMR API module makes sure there is an "Unknown" provider in the system, and in our case it's overwriting the "admin" provider set up by ref app. Will keep you posted, but thanks for the help!

Fixed!  This was the issue:  https://openmrs.atlassian.net/browse/RA-2061

Anyway, this is ready for merge, but I'm finishing up for the day shortly and I'm on vacation the first half of next week... so the smart thing is for me not to merge this until I'm back... :)

**Review by denniskigen — APPROVED** [review]
LGTM. Thanks, @mogoodrich!

**Comment by mogoodrich** [comment]
Thanks @denniskigen  @ibacher !  Will test out tomorrow how this is looking on dev3...

**Comment by vasharma05** [comment]
Hi @mogoodrich!
Can you also configure the beds for a location, or if it's already configured, can you please share the location.
Thanks for the great work!

**Pull request merged**